# **🚀 Факторизация и другие интересные операции в Julia**  
# **Урок Тринадцатый.**

По мотивам работы Андреаса Ноака Йенсена (MIT & JuliaComputing) [(Twitter)](https://x.com/anoackjensen)  
(с авторскими дополнениями Сергея Соболевского)

<br>

##  **📌 Темы:**  

1. **Что такое факторизация матрицы**
   
2. **Основные виды факторизации матриц.**
   
   2.1 **LU-факторизация**

   2.2 **QR-факторизация**

   2.3 **Cholesky-факторизация**

   2.4 **Сингулярное разложение (SVD)**

   2.5 **Разложение по собственным значениям и собственным векторам (EVD)**

   
3. **Специальные матричные структуры.**
   
4. **Общая линейная алгебра.**



<br>


<br>


<br>

---

<br>

# 📚 **1. Что такое факторизация матрицы?**

**Факторизация матрицы** (или разложение матрицы) — это представление исходной матрицы в виде произведения нескольких матриц определённого типа, обычно более простых по структуре. 

Суть метода — исходную задачу (например, решение системы линейных уравнений, нахождение определителя или обратной матрицы) привести к более простой и численно устойчивой форме.

<br>


### 🧮 Для чего нужна факторизация?

Факторизация применяется в различных областях математики и вычислительной науки:

- **Решение систем линейных уравнений**: упрощает решение, особенно при множестве задач с одинаковой матрицей, но разными правыми частями.
- **Нахождение обратных матриц**: облегчает и ускоряет процесс вычисления.
- **Определение ранга матрицы и её структуры** (например, собственных значений и векторов).
- **Упрощение вычислений** в численных методах, повышая устойчивость и точность расчётов.

<br>



<br>

<br>


### ⚙️ Настройка нашей среды **Julia**.

Прежде чем начать, давайте настроим нашу среду , а заодно повторим как подключить в **Julia** библиотеку `LinearAlgebra`, чтобы дальше мы могли применять факторизации и работать со специальными структурами матриц.  

- **Подключаем пакет `LinearAlgebra` :**

In [4]:
using LinearAlgebra # Подключаем пакет LinearAlgebra



- **Создаем случайную матрицу $A$ (3x3) :**

In [ ]:
A = rand(3, 3)  # Создаем случайную матрицу 3x3

- **Создаем вектор $x$ из трех единиц :**

In [ ]:
x = fill(1, (3,))  # Создаем вектор из трех 1-единиц

- **Вычисляем для пробы результат умножения матрицы $A$ на вектор $x$ :**

In [ ]:
b = A * x  # Вычисляем результат умножения матрицы A на вектор x

**🚩 "Бинго!" Всё получилось, идём дальше!**

<br>

---

<br>

## **🔢 2. Основные виды факторизации матриц.**

<br>
Наиболее часто применяются следующие виды разложений:

1. **LU-факторизация**
   
2. **QR-факторизация**
   
3. **Cholesky-факторизация**
   
4. **Сингулярное разложение (SVD)**
   
5. **Разложение по собственным значениям и собственным векторам (EVD)**

<br>

Давайте рассмотрим кратко каждый этот тип подробнее, но вначале настроим нашу среду **Julia** для работы.

<br>

---

<br>

### 🟢 **2.1 LU-факторизация**

<br>

**LU-факторизация** — это представление матрицы $A$ в виде произведения нижней треугольной матрицы $L$ (Lower triangular) и верхней треугольной матрицы $U$ (Upper triangular):

$$
A = L \cdot U
$$

Чтобы повысить численную устойчивость вычислений, обычно используется матрица перестановок $P$:

$$
P \cdot A = L \cdot U
$$

Здесь:

- $P$ — матрица перестановок (переставляет строки исходной матрицы $A$),
- $L$ — нижняя треугольная матрица с единичными элементами на главной диагонали,
- $U$ — верхняя треугольная матрица.

<br>

#### 🔹 **Применение LU-факторизации:**

- Решение систем линейных уравнений.
- Вычисление определителей матриц.
- Нахождение обратной матрицы.


В языке Julia LU-факторизация матрицы $A$ выполняется с помощью функции `lu()`:

In [5]:
Alu = lu(A)  # Разложение матрицы A на L и U

LU{Float64, Matrix{Float64}, Vector{Int64}}
L factor:
2×2 Matrix{Float64}:
 1.0  0.0
 0.5  1.0
U factor:
2×2 Matrix{Float64}:
 4.0  5.0
 0.0  0.5

👉 Julia автоматически создаёт объект специального составного типа, который хранит компоненты факторизации.

Извлечь компоненты факторизации можно с помощью специальных полей объекта:

- **Матрица перестановок:**

In [ ]:
P = Alu.P

- **Нижняя треугольная матрица $L$ :**

In [ ]:
L = Alu.L

- **Верхняя треугольная матрица $U$ :**

In [ ]:
U = Alu.U

### 🗒 **Примеры использования LU-факторизации:**

#### 🔹 **Пример 1: Решение системы линейных уравнений:**

Пусть задана система линейных уравнений:

$$
\begin{cases}
2x + 3y = 5 \\
4x + 5y = 11
\end{cases}
$$

Запишем в матричной форме:

In [6]:
A = [2 3; 4 5]
b = [5; 11]

2-element Vector{Int64}:
  5
 11

In [7]:
Alu = lu(A) # Разложение матрицы A на L и U

LU{Float64, Matrix{Float64}, Vector{Int64}}
L factor:
2×2 Matrix{Float64}:
 1.0  0.0
 0.5  1.0
U factor:
2×2 Matrix{Float64}:
 4.0  5.0
 0.0  0.5

- **Решение через исходную матрицу:**

In [8]:
x_direct = A \ b  # Решение системы уравнений Ax = b с помощью обратного хода

2-element Vector{Float64}:
  4.0
 -1.0

- **Решение через факторизацию LU:**

In [9]:
x_lu = Alu \ b  # Решение системы уравнений Ax = b с помощью LU-разложения

2-element Vector{Float64}:
  4.0
 -1.0

**Убеждаемся, что оба метода дают одинаковый результат.**


#### 🔹 **Пример 2: Вычисление определителя:**

Определитель матрицы можно вычислить напрямую или используя LU-факторизацию:

In [11]:
det_A_direct = det(A) # Вычисление определителя матрицы A с помощью обратного хода

-2.0

In [ ]:

det_A_lu = det(Alu) # Вычисление определителя матрицы A с помощью LU-разложения

-2.0

Результаты будут близки до численной погрешности:

In [12]:
det(A) ≈ det(Alu) # Проверка равенства определителей

true

#### 👉 **Преимущества LU-факторизации в Julia:**

- **Численная стабильность**: Благодаря матрице перестановок факторизация становится устойчивой.
  
- **Эффективность**: LU-факторизацию можно использовать повторно для различных вычислений (решение систем, нахождение обратной матрицы, вычисление определителей).



<br>

---

<br>


<br>

### 🟢 **2.2 QR-факторизация**

**QR-факторизация** — это разложение матрицы $ A $ в виде произведения двух специальных матриц: ортогональной (унитарной) матрицы $ Q $ и верхней треугольной матрицы $ R $ :

<br>

$$
A = Q \cdot R
$$

**Характеристики матриц разложения:**

- Матрица $ Q $ — ортогональная (унитарная), то есть выполняется:
$$
Q^T Q = I
$$

Это означает, что столбцы матрицы $ Q $ :

- **ортогональны** друг другу (их скалярное произведение равно нулю);
- имеют **единичную длину** (нормированы).

- Матрица $ R $ — верхняя треугольная, то есть все её элементы ниже главной диагонали равны нулю.


<br>

#### 📌 **Применение QR-факторизации:**

QR-факторизация широко используется в вычислительной математике, статистике и инженерии:

- **Решение задач наименьших квадратов** (линейная регрессия).
  
- **Решение линейных систем уравнений** (особенно устойчивых к численным погрешностям).
  
- **Вычисление собственных значений и векторов** (QR-алгоритм).

<br>

### 🗒  **Примеры выполнения QR-факторизации в Julia:**

#### 🔹 **Пример 1.**

В языке Julia **QR-факторизацию** матрицы $ A $ можно выполнить с помощью функции `qr()`:


In [13]:
Aqr = qr(A) # Разложение матрицы A на Q и R

LinearAlgebra.QRCompactWY{Float64, Matrix{Float64}, Matrix{Float64}}
Q factor: 2×2 LinearAlgebra.QRCompactWYQ{Float64, Matrix{Float64}, Matrix{Float64}}
R factor:
2×2 Matrix{Float64}:
 -4.47214  -5.81378
  0.0      -0.447214


После этого результат факторизации сохраняется в специальном объекте, из которого легко извлечь матрицы $ Q $ и $ R $ :

- **Матрица $ Q $** (ортогональная матрица):

In [14]:
Q = Aqr.Q

2×2 LinearAlgebra.QRCompactWYQ{Float64, Matrix{Float64}, Matrix{Float64}}

- **Матрица $ R $** (верхняя треугольная матрица):

In [15]:
R = Aqr.R

2×2 Matrix{Float64}:
 -4.47214  -5.81378
  0.0      -0.447214

👉 Таким образом, после выполнения **QR-факторизации** мы можем работать с её компонентами напрямую и эффективно.

#### 🔹 **Пример 2.**

**Краткий пример использования, :) для закрепления темы:**

Рассмотрим простую матрицу:

```julia
A = [1 2; 
     3 4; 
     5 6]
```

In [16]:
A = [1 2; 3 4; 5 6]

3×2 Matrix{Int64}:
 1  2
 3  4
 5  6

In [17]:
Aqr = qr(A) # Разложение матрицы A на Q и R

LinearAlgebra.QRCompactWY{Float64, Matrix{Float64}, Matrix{Float64}}
Q factor: 3×3 LinearAlgebra.QRCompactWYQ{Float64, Matrix{Float64}, Matrix{Float64}}
R factor:
2×2 Matrix{Float64}:
 -5.91608  -7.43736
  0.0       0.828079

In [18]:
Q = Aqr.Q   # ортогональная матрица

3×3 LinearAlgebra.QRCompactWYQ{Float64, Matrix{Float64}, Matrix{Float64}}

In [19]:
R = Aqr.R   # верхняя треугольная матрица

2×2 Matrix{Float64}:
 -5.91608  -7.43736
  0.0       0.828079

👉 **Теперь исходная матрица $ A $ может быть представлена в виде:**

$$
A = Q \cdot R
$$


In [ ]:
A = Q * R  # Проверка разложения

<br>

#### 👉 **Преимущества использования QR-факторизации:**

- **Численная устойчивость**: QR-факторизация обычно более стабильна и точна, чем другие методы (например, LU-факторизация), особенно при работе с плохо обусловленными матрицами.
  
- **Эффективность**: Позволяет быстро и устойчиво решать системы линейных уравнений и задачи регрессии.



<br>

---

<br>

<br>

### 🟢 **2.3 Cholesky-факторизация**

**Cholesky-факторизация** применяется исключительно для **симметричных и положительно определённых** матриц. Она представляется в виде произведения нижней треугольной матрицы $L$ и её транспонированной матрицы $L^T$:

$$
A = L \cdot L^T
$$

где:

- $L$ — нижняя треугольная матрица (все элементы выше диагонали равны нулю).
- $L^T$ — верхняя треугольная матрица, транспонированная к $L$.

<br>

##### В Julia **Cholesky-факторизация** выполняется функцией `cholesky()` .

<br>

### 🔹 **Особенности Cholesky-факторизации:**

- Используется только для симметричных и положительно определённых матриц.
- Является более простым и численно стабильным методом по сравнению с LU-факторизацией.
- Более эффективна с точки зрения производительности, так как требует примерно в два раза меньше вычислений, чем LU-факторизация.

<br>


#### 📌 **Применение Cholesky-факторизации:**

- Решение систем линейных уравнений (особенно с симметричными матрицами).
  
- Вычисления в статистике (например, линейные регрессионные модели, многомерное нормальное распределение).
  
- Численные методы оптимизации и анализа данных.

<br>


### 🗒  **Примеры Cholesky-факторизации в Julia:**

#### 🔹 **Пример 1.**

1. **Определяем симметричную положительно определённую матрицу $A$:**

```julia
A = [
    25.0  15.0  -5.0;
    15.0  18.0   0.0;
    -5.0   0.0  11.0
]
```


In [22]:
A = [
    25.0  15.0  -5.0;
    15.0  18.0   0.0;
    -5.0   0.0  11.0
]

3×3 Matrix{Float64}:
 25.0  15.0  -5.0
 15.0  18.0   0.0
 -5.0   0.0  11.0

2. **Выполняем проверку симметричности :**

In [23]:
println("Is A symmetric? ", issymmetric(A)) # Проверка симметричности матрицы A

Is A symmetric? true


3. **Выполнение Cholesky-факторизации :**

In [24]:
cholesky_factor = cholesky(A) # Разложение матрицы A на L и L^T

Cholesky{Float64, Matrix{Float64}}
U factor:
3×3 UpperTriangular{Float64, Matrix{Float64}}:
 5.0  3.0  -1.0
  ⋅   3.0   1.0
  ⋅    ⋅    3.0

4. **Выводим результат, т.е. извлекаем компоненты факторизации следующим образом:**

- **Нижняя треугольная матрица $L$ :**

In [26]:
L = cholesky_factor.L # нижняя треугольная матрица

3×3 LowerTriangular{Float64, Matrix{Float64}}:
  5.0   ⋅    ⋅ 
  3.0  3.0   ⋅ 
 -1.0  1.0  3.0

In [25]:
println("Cholesky factor L:\n", cholesky_factor.L) # Матрица L

Cholesky factor L:
[5.0 0.0 0.0; 3.0 3.0 0.0; -1.0 1.0 3.0]


#### 🔹 **Пример 2. Использования Cholesky-факторизации, ещё один:** :*

Рассмотрим симметричную и положительно определённую матрицу:

In [27]:
A = [4 2; 2 3]

2×2 Matrix{Int64}:
 4  2
 2  3

- **Проверяем положительную определённость (собственные значения положительны):**

In [28]:
eigvals(A) # Вычисление собственных значений матрицы A

2-element Vector{Float64}:
 1.43844718719117
 5.561552812808831


- **Выполняем Cholesky-факторизацию:**

In [32]:
A_chol = cholesky(A)

Cholesky{Float64, Matrix{Float64}}
U factor:
2×2 UpperTriangular{Float64, Matrix{Float64}}:
 2.0  1.0
  ⋅   1.41421

- **Извлекаем нижнюю треугольную матрицу L :**

In [33]:
L = A_chol.L

2×2 LowerTriangular{Float64, Matrix{Float64}}:
 2.0   ⋅ 
 1.0  1.41421

- **Проверим правильность разложения :**

In [34]:
A_reconstructed = L * L' # Восстановление матрицы A

2×2 Matrix{Float64}:
 4.0  2.0
 2.0  3.0

- **Проверка равенства определителей :**

In [35]:
det(A_reconstructed) ≈ det(A) # Проверка равенства определителей

true

#### 👉 **Преимущества Cholesky-факторизации:**

- Высокая численная устойчивость.
  
- Минимальные требования к памяти.
  
- Повышенная производительность для специальных типов матриц.

<br>

🚀 Используя **Cholesky-факторизацию** в Julia, можно эффективно и стабильно решать системы линейных уравнений, часто возникающие в прикладных задачах статистики и численного анализа.

<br>

---

<br>

### 🟢 **2.4 Сингулярное разложение (SVD, Singular Value Decomposition)**

**Сингулярное разложение (SVD)** — это факторизация любой матрицы $A$ (включая неквадратные) в виде произведения трёх матриц:

<br>

$$
A = U \cdot \Sigma \cdot V^T
$$

**где:**

- $U$ — ортогональная матрица размера $m \times m$, содержащая левые сингулярные векторы.
  
- $V$ — ортогональная матрица размера $n \times n$, содержащая правые сингулярные векторы.
  
- $\Sigma$ — прямоугольная диагональная матрица размера $m \times n$, содержащая сингулярные числа на диагонали (упорядочены по убыванию и всегда неотрицательные).
  
<br>

##### **В Julia сингулярное разложение выполняется функцией `svd()`.**

<br>

#### 🔹 **Особенности сингулярного разложения (SVD):**

- Работает для любой матрицы (квадратной и неквадратной).
  
- Диагональные элементы (сингулярные числа) матрицы $\Sigma$ неотрицательны и расположены по убыванию.
  
- Позволяет оценивать ранг матрицы и её численную устойчивость.

<br>

#### 📌 **Применение SVD:**

- **Сжатие и обработка изображений:** Уменьшение размерности данных, PCA (метод главных компонент).
  
- **Анализ данных и машинное обучение:** Рекомендательные системы, кластеризация, обработка естественного языка.
  
- **Определение ранга матрицы:** Оценка численной стабильности и качества аппроксимации данных.
  
  <br>

### 🗒  **Примеры SVD-разложения в Julia:**

#### 🔹 **Пример 1.**

1. **Определяем симметричную положительно определённую матрицу $A$:**

```julia
A = [
    25.0  15.0  -5.0;
    15.0  18.0   0.0;
    -5.0   0.0  11.0
]
```

In [ ]:
A = [
    25.0  15.0  -5.0;
    15.0  18.0   0.0;
    -5.0   0.0  11.0
]

2. **Выполним сингулярное разложение при помощи функцией `svd()`:**

In [ ]:
Asvd = svd(A)

3. **Полученные компоненты факторизации извлекаются следующим образом:**
   


- Матрица $U$ (левые сингулярные векторы):

In [ ]:
U = Asvd.U

- Диагональная матрица сингулярных чисел $\Sigma$:

In [ ]:
S = Asvd.S

- Матрица $V$ (правые сингулярные векторы):

In [ ]:
V = Asvd.V

- **Восстановление исходной матрицы по её SVD-разложению :**

A_reconstructed = U * Diagonal(S) * V'

### 🔹 **Пример 2 использования SVD:**

Рассмотрим новую матрицу:

```julia
A = [3 1; 1 3; 1 1]
```



In [ ]:
A = [3 1; 1 3; 1 1]

- **Выполняем SVD-разложение :**

In [ ]:
Asvd = svd(A) # Сингулярное разложение матрицы A

- **Матрица $U$ (левые сингулярные векторы):**

In [ ]:
U = Asvd.U # ортогональная матрица

- **Диагональная матрица сингулярных чисел $\Sigma$ :**

In [ ]:
S = Asvd.S # диагональная матрица

- **Матрица $V$ (правые сингулярные векторы):**

In [ ]:
V = Asvd.V # ортогональная матрица

- **Восстановление исходной матрицы по её SVD-разложению:**

In [ ]:
A_reconstructed = U * Diagonal(S) * V' # Восстановление матрицы A

#### 👉 **Преимущества в использовании SVD-разложения:**

- Позволяет работать с любыми матрицами, включая неквадратные.
  
- Высокая численная устойчивость и надёжность.
  
- Эффективно решает задачи, связанные с обработкой больших объёмов данных, особенно при анализе данных и машинном обучении.

<br>

<br>

---

<br>

### 🟢 **2.5 Собственные разложения (Eigendecompositions -EVD)**

<br>

**Собственное разложение (Eigendecomposition)** — это представление её в виде произведения трёх матриц:

<br>

$$
A = V \Lambda V^{-1},
$$


**где:**

- $ V $ — матрица собственных векторов,
- $ \Lambda $ — диагональная матрица собственных значений.


### 🗒  **Примеры EVD-разложения в Julia:**

#### 🔹 **Пример 1.**

1. **Предположим, что у нас есть симметричная матрица $A$, созданная следующим образом :**


```julia
A = [
    25.0  15.0  -5.0;
    15.0  18.0   0.0;
    -5.0   0.0  11.0
]
```

In [8]:
A = [
    25.0  15.0  -5.0;
    15.0  18.0   0.0;
    -5.0   0.0  11.0
]

3×3 Matrix{Float64}:
 25.0  15.0  -5.0
 15.0  18.0   0.0
 -5.0   0.0  11.0

- **Или мы можем создать симметричную матрицу:**

In [ ]:
Asym = A + A'# Создание симметричной матрицы

- **Чтобы вычислить собственное разложение, используем функцию `eigen`:**

In [11]:
AsymEig = eigen(Asym) # Вычисление собственных значений и векторов матрицы Asym

Eigen{Float64, Float64, Matrix{Float64}, Vector{Float64}}
values:
3-element Vector{Float64}:
  8.990924996526852
 24.03136729085444
 74.9777077126187
vectors:
3×3 Matrix{Float64}:
  0.595003  -0.178127   0.783736
 -0.660893   0.446485   0.603219
  0.457375   0.876882  -0.147937

👉 **Полученный нами результат вычисления представляет собой собственные значения и собственные векторы симметричной матрицы:**

- **Собственные значения (eigenvalues)**:
```julia
[8.990924996526852, 24.03136729085444, 74.9777077126187]
```
Собственные значения отражают, как исходная матрица масштабирует (растягивает или сжимает) пространство вдоль направлений, заданных соответствующими собственными векторами. 

Чем больше собственное значение, тем сильнее растяжение в направлении его собственного вектора.

<br>

- **Собственные векторы (eigenvectors)**:

Это столбцы (векторы) нашей матрицы –  направления (нормализованные векторы длины 1), по которым происходит масштабирование пространства. Каждый собственный вектор соответствует собственному значению с тем же индексом:

- Вектор 1 `[0.595003, -0.660893, 0.457375]` соответствует собственному значению `8.9909`.
  
- Вектор 2`[-0.178127, 0.446485, 0.876882]` соответствует собственному значению `24.0314`.
  
- Вектор 3 `[0.783736, 0.603219, -0.147937]` соответствует собственному значению `74.9777`.

<br>

>**Говоря простыми словами наша исходная матрица при умножении на данные векторы просто меняет их длину, не меняя направления.**
>**Эти три направления ортогональны (перпендикулярны друг другу), так как исходная матрица симметрична.**
>**Таким образом, полученный нами результат — не что инное как полное разложение матрицы на ортогональные направления и коэффициенты масштабирования.**

<br>

#### 🔹 Рассмотрим доступ к собственным значениям и векторам матрицы:


- **Задаем исходную матрицу A, которая будет использована для демонстрации собственных значений и векторов**

```julia
A = [
    25.0  15.0  -5.0;
    15.0  18.0   0.0;
    -5.0   0.0  11.0
]
```

A = [
    25.0  15.0  -5.0;
    15.0  18.0   0.0;
    -5.0   0.0  11.0
]

- **Создаём симметричную матрицу.**

Симметричная матрица важна для вычисления собственных значений и векторов, так как обеспечивает реальные значения и ортогональность собственных векторов.

In [ ]:
Asym = A + A'

- **Выполняем EVD-разложение**

Вычисляем собственные значения и собственные векторы, получая EVD-разложение матрицы.

In [ ]:
AsymEig = eigen(Asym)

- **Получаем доступ к собственным значениям и векторам:**
  
  

**Собственные значения** - показывают степень масштабирования пространства по соответствующим направлениям (собственным векторам):

In [ ]:
λ = AsymEig.values                     # Собственные значения
println("Собственные значения: ", λ)   # Вывод собственных значений

**Собственные векторы** - задают направления, по которым происходит масштабирование:

In [ ]:
V = AsymEig.vectors                    # Собственные векторы
println("Собственные векторы: \n", V)  # Вывод собственных векторов

- **Проверим корректность факторизации**
  
А именно, что произведение `Asym` на `V` дает то же самое, что и произведение `V` на диагональную матрицу собственных значений

In [ ]:
println("Проверка разложения: ", norm(Asym * V - V * Diagonal(λ)))  # должно быть ≈ 0

<br>

<br>


#### 🔹 Использование EVD-факторизации для оптимизации вычислений.

**Используя EVD-факторизацию, обратная матрица вычисляется намного эффективнее, особенно для больших матриц**

In [12]:
Asym_inv_eig = inv(AsymEig)

3×3 Matrix{Float64}:
  0.0488889  -0.0407407   0.0222222
 -0.0407407   0.0617284  -0.0185185
  0.0222222  -0.0185185   0.0555556

**Проверим точность вычисления обратной матрицы :**

In [13]:
println("Эффективно вычисленная обратная матрица через EVD:\n", Asym_inv_eig)

Эффективно вычисленная обратная матрица через EVD:
[0.048888888888888864 -0.04074074074074075 0.022222222222222202; -0.04074074074074074 0.06172839506172842 -0.01851851851851851; 0.022222222222222185 -0.0185185185185185 0.055555555555555594]


**Убедимся в корректности (должна получиться единичная матрица)**

Этим шагом мы проверяем, что при умножении исходной матрицы на обратную, получается единичная матрица (I):

In [14]:
println("Проверка корректности вычисления обратной матрицы:\n", Asym_inv_eig * Asym)

Проверка корректности вычисления обратной матрицы:
[0.9999999999999988 -1.1102230246251565e-15 -2.220446049250313e-16; 8.881784197001252e-16 1.000000000000001 1.6653345369377348e-16; -1.7763568394002505e-15 -4.440892098500626e-16 1.000000000000001]


**Сравним со стандартным способом вычисления обратной матрицы, что менее эффективно. Но так ли это?**

Проверим, насколько сильно отличается результат, полученный через стандартный подход

In [15]:
Asym_inv_standard = inv(Asym)
println("Разница между двумя способами вычисления обратной матрицы: ", norm(Asym_inv_eig - Asym_inv_standard))

Разница между двумя способами вычисления обратной матрицы: 7.28583859910259e-17


#### 🔹 **Выводы:**

Использование специальных типов факторизаций в Julia позволяет:

👉 Хранить результаты вычислений компактно и логично.
  
👉 Быстро извлекать значения и векторы для дальнейших вычислений.
  
👉 Автоматически использовать оптимизированные методы, учитывающие структуру факторизации.


<br>

<br>

---

<br>

### 🟢 **2.6 Другие менее распространённые факторизации**

Помимо наиболее распространённых факторизаций, уже рассмотренных нами выше, существуют также менее известные, но важные методы разложения матриц, которые применяются в специфических и продвинутых задачах линейной алгебры и численного анализа.

<br>

### 📌 **Разложение Шура (Schur Decomposition)**

<br>

**Разложение Шура** представляет собой представление квадратной матрицы $A$ в виде произведения:

<br>

$$
A = Q \cdot T \cdot Q^*
$$

**где:**

- $Q$ — унитарная (ортогональная) матрица, то есть выполняется условие $Q^* Q = I$;
  
- $T$ — верхняя треугольная (для комплексной матрицы) или верхняя квазитреугольная матрица (для вещественной матрицы, известная как **Real Schur form**).

<br>


**Важно:**
>Однако, разработчики языка **Julia** приняли решение называть при разложении Шура унитарную матрицу в объекте не $Q$, а $Z$, чтобы избежать путаницы с другими разложениями, например **QR-разложением**, где уже используется буква $Q$. Таким образом, для ясности, это чисто техническое решение, сделанное разработчиками **Julia**.

<br>

#### 🔹 **Использование разложения Шура:**

- Удобно для вычисления собственных значений (на диагонали матрицы $T$).
  
- Применяется в теории управления и численном анализе.
  
- Эффективно используется при реализации QR-алгоритма нахождения собственных значений.

<br>

### 🗒  **Пример использования разложения Шура в Julia:**

#### 🔹 **Пример 1.**

- **Определяем матрицу $A$:**

```julia 
A = [
     4  3; 
    -5 -2
    ]
```

In [16]:
A = [4 3; -5 -2] # Создание матрицы A

2×2 Matrix{Int64}:
  4   3
 -5  -2

- **Выполняем разложение Шура (Schur decomposition) матрицы**:

In [17]:
SchurF = schur(A) # Получение матрицы Шура

Schur{Float64, Matrix{Float64}, Vector{ComplexF64}}
T factor:
2×2 Matrix{Float64}:
  1.0      0.837722
 -7.16228  1.0
Z factor:
2×2 Matrix{Float64}:
 0.811242  -0.58471
 0.58471    0.811242
eigenvalues:
2-element Vector{ComplexF64}:
 0.9999999999999999 + 2.449489742783178im
 0.9999999999999999 - 2.449489742783178im

**Получаем компоненты факторизации**

Мы помним, что разработчики языка Julia приняли решение называть унитарную матрицу в этом объекте не $Q$, а $Z$, чтобы избежать путаницы с другими разложениями.

- $Q$ – унитарная (ортогональная для вещественных матриц) матрица:

In [19]:
Q = SchurF.Z # Получение матрицы Q

2×2 Matrix{Float64}:
 0.811242  -0.58471
 0.58471    0.811242

- $𝑇$ – верхняя квазитреугольная матрица Шура (для вещественных матриц блоками 1x1 или 2x2)

In [20]:
T = SchurF.T # Получение верхнетреугольной матрицы T

2×2 Matrix{Float64}:
  1.0      0.837722
 -7.16228  1.0

👉 **Матрица $T$ имеет верхнетреугольную форму, что облегчает чтение собственных значений непосредственно с диагонали.**

<br>

---

<br>

### 📌 **Разложение Жордана (Jordan Normal Form)**

<br>

**Разложение Жордана** представляет собой представление квадратной матрицы $A$ в виде подобия с блочно-диагональной матрицей:

<br>

$$
A = P \cdot J \cdot P^{-1}
$$
**где:**

- $P$ - **Матрица перехода** - содержит в качестве столбцов собственные и обобщенные (присоединенные) векторы. Используется для перехода в новый базис, в котором исходная матрица имеет максимально простой вид (Жорданова форма).
  
- $J$ -**Матрица Жордана** - это верхнетреугольная матрица, содержащая собственные значения на диагонали и единицы над диагональю, если собственных векторов недостаточно для диагонализации.

<br>

>**Разложение применяется, когда матрица не диагонализируема, то есть ей не хватает полного набора собственных векторов.**

<br>

#### 🔹 **Использование разложения Жордана:**

- Теоретический анализ свойств матриц и операторов.
  
- Решение дифференциальных и разностных уравнений.
  
- Изучение структуры линейных операторов.

<br> 

📚 В Julia нет встроенной функции для прямого вычисления **разложения Жордана (Jordan decomposition)**, поскольку это разложение неустойчиво при численных вычислениях (чувствительно к погрешностям).

**Поэтому используют дополнительные пакеты:**
- Для численных расчетов лучше всего использовать:
   - `JordanForm.jl`
   - `LinearAlgebraX.jl`
- Для символьных расчетов рекомендуется:
   - `Symbolics.jl`
  
<br>

### 🗒  **Пример численного вычисления с помощью пакета `JordanForm.jl`**

#### 🔹 **Пример 1.**


**Установим пакет `JordanForm.jl`:**

In [ ]:
using Pkg
Pkg.add("JordanForm")

   Resolving package versions...
   Installed HypergeometricFunctions ─── v0.3.27
   Installed PDMats ──────────────────── v0.11.32
   Installed StatsFuns ───────────────── v1.3.2
   Installed TimerOutputs ────────────── v0.5.28
   Installed JordanForm ──────────────── v0.1.0
   Installed Accessors ───────────────── v0.1.42
   Installed MutableArithmetics ──────── v1.6.4
   Installed DomainSets ──────────────── v0.7.15
   Installed LabelledArrays ──────────── v1.16.0
   Installed EnumX ───────────────────── v1.0.4
   Installed FunctionWrappers ────────── v1.1.3
   Installed MultivariatePolynomials ─── v0.5.7
   Installed ArnoldiMethod ───────────── v0.4.0
   Installed RecursiveArrayTools ─────── v3.31.0
   Installed Bijections ──────────────── v0.1.9
   Installed SciMLBase ───────────────── v2.75.1
   Installed IfElse ──────────────────── v0.1.1
   Installed CompositeTypes ──────────── v0.1.4
   Installed ADTypes ─────────────────── v1.14.0
   Installed Rmath_jll ───────────────── v0.5

- **Производим подключение пакета JordanForm для вычисления разложения Жордана**

In [ ]:
using JordanForm

- **Задаем исходную матрицу $A$, для которой хотим найти Жорданову форму:**

```julia
A = [6 2 1; 
     0 3 0; 
     0 0 3]
```


In [ ]:
A = [6 2 1; 0 3 0; 0 0 3] # Создание матрицы A

3×3 Matrix{Int64}:
 6  2  1
 0  3  0
 0  0  3

- **Вычисляем Жорданову нормальную форму $J$ и матрицу перехода $P$:**

In [ ]:
J, P = jordan_form(A) # Получение жордановой формы матрицы A

JordanFactorization{ComplexF64, ComplexF64}(ComplexF64[-0.6666666666666666 + 0.0im -0.3333333333333333 + 0.0im 1.0 + 0.0im; 1.0 + 0.0im 0.0 + 0.0im -0.0 - 0.0im; 0.0 + 0.0im 1.0 + 0.0im -0.0 - 0.0im], ComplexF64[3.0 + 0.0im 0.0 + 0.0im 0.0 + 0.0im; 0.0 + 0.0im 3.0 + 0.0im 0.0 + 0.0im; 0.0 + 0.0im 0.0 + 0.0im 6.0 + 0.0im])

- **Выводим полученную Жорданову форму матрицы $A$ :**

In [29]:
println("Жорданова форма J:\n", J) # Вывод жордановой формы

Жорданова форма J:
ComplexF64[-0.6666666666666666 + 0.0im -0.3333333333333333 + 0.0im 1.0 + 0.0im; 1.0 + 0.0im 0.0 + 0.0im -0.0 - 0.0im; 0.0 + 0.0im 1.0 + 0.0im -0.0 - 0.0im]


- **Выводим матрицу перехода (базиса), состоящую из собственных и присоединенных векторов :**

In [30]:
println("Матрица перехода P:\n", P) # Вывод матрицы перехода

Матрица перехода P:
ComplexF64[3.0 + 0.0im 0.0 + 0.0im 0.0 + 0.0im; 0.0 + 0.0im 3.0 + 0.0im 0.0 + 0.0im; 0.0 + 0.0im 0.0 + 0.0im 6.0 + 0.0im]


- **Проверяем корректность полученного разложения, вычисляя разницу между исходной матрицей и произведением матриц разложения (должна быть очень близка к 0) :**

In [31]:
println("Проверка корректности разложения: ", norm(A - P*J*inv(P))) # должно быть ≈ 0

Проверка корректности разложения: 8.552127740444998


❌ **Полученный результат означает, что разложение некорректно, так как оно должно быть очень близким к нулю (например, порядка 1e-12 или меньше).**

Это связано с тем, что алгоритмы численного вычисления Жордановой формы являются неустойчивыми и могут давать существенные ошибки из-за погрешностей округления. Именно поэтому вычисление Жордановой формы практически всегда рекомендуют выполнять:

- либо аналитически (символьно),
  
- либо для матриц, которые специально подобраны и не дают ошибок округления.

<br>

**Как же исправить эту проблему?**

:) Используйте символьное вычисление !!!

<br>

### ✅ **Символьное вычисление с пакетом `Symbolics.jl`**



**Использование символьных вычислений**

- Позволяет нам получить аналитические формулы.
  
- Удобно для теоретических и исследовательских задач, когда числовых значений элементов нет или они неизвестны заранее.
  
- Упрощает решение общих задач линейной алгебры и матричного анализа.

<br>

**Установка:**

```julia
Pkg.add("Symbolics")
Pkg.add("SymbolicUtils")
```

#### 🔹 **Пример 2.**

Корректный пример с символикой:

**Выполним установку пакета `Symbolics.js` в Julia:**

In [45]:
using Pkg
Pkg.add("Symbolics")
Pkg.add("SymbolicUtils")

   Resolving package versions...
  No Changes to `~\Introduction-to-Julia\Project.toml`
  No Changes to `~\Introduction-to-Julia\Manifest.toml`
   Resolving package versions...
    Updating `~\Introduction-to-Julia\Project.toml`
⌅ [d1185830] + SymbolicUtils v1.7.1
  No Changes to `~\Introduction-to-Julia\Manifest.toml`


In [35]:
using Symbolics, LinearAlgebra

**Определяем матрицу $A$:**

In [36]:
A = [6 2 1; 0 3 0; 0 0 3] # Создание матрицы A

3×3 Matrix{Int64}:
 6  2  1
 0  3  0
 0  0  3

**Производим символьное вычисление собственных значений:**

Используем функцию `eigvals(A)`

In [38]:
eigvals_A = eigvals(A) # Вычисление собственных значений матрицы A

3-element Vector{Float64}:
 3.0
 3.0
 6.0

In [39]:
println("Собственные значения:\n", eigvals_A) # Вывод собственных значений

Собственные значения:
[3.0, 3.0, 6.0]


<br>

#### 🔹 **Пример 3.**

На этом примере продемонстрируем использование символьных вычислений в Julia с помощью пакета Symbolics.jl. Убедимся, что символьные вычисления помогают аналитически исследовать матрицы, позволяя работать не только с числовыми, но и с переменными элементами матриц.

**Мы зададим символьные переменные, которые затем можно использовать в символьной матрице. Это означает, что матрица может содержать не числа, а переменные, полезные при аналитических расчётах.**

1. Задаём **символьные переменные** (элементы матрицы):

In [56]:
@variables a11 a12 a13 a21 a22 a23 a31 a32 a33 # Создание символьных переменных

9-element Vector{Num}:
 a11
 a12
 a13
 a21
 a22
 a23
 a31
 a32
 a33

 2. Определение **символьной матрицы**, которая состоит полностью из символьных элементов. Её можно использовать для вывода общих формул и аналитического исследования свойств матриц, таких как собственные значения и векторы в общем виде.

In [64]:
A = [a11 a12 a13; 
     a21 a22 a23; 
     a31 a32 a33]

3×3 Matrix{Num}:
 a11  a12  a13
 a21  a22  a23
 a31  a32  a33

3. Получаем **характеристический полином** (аналитически!):

In [65]:
λ = Symbolics.scalarize(@variables λ)[1] # Создание символьной переменной λ
char_poly = det(A - λ * I(3)) # Вычисление характеристического полинома

(a11 - λ)*((a22 - λ)*(a33 - λ) - a23*a32) - a12*(a21*(a33 - λ) - a23*a31) + a13*(a21*a32 - (a22 - λ)*a31)

In [66]:
println("Характеристический полином матрицы A:\n", char_poly) # Вывод характеристического полинома

Характеристический полином матрицы A:
(a11 - λ)*((a22 - λ)*(a33 - λ) - a23*a32) - a12*(a21*(a33 - λ) - a23*a31) + a13*(a21*a32 - (a22 - λ)*a31)


4. Затем приводится конкретная числовая матрица, для которой будут вычислены собственные значения. Из этого видно, как можно использовать символьные методы и подходы и для конкретных числовых матриц.

In [67]:
B = [6 2 1; 
     0 3 0; 
     0 0 3] # Создание матрицы B

3×3 Matrix{Int64}:
 6  2  1
 0  3  0
 0  0  3

5. Вычисляем **характеристический полином** конкретной матрицы $B$:

In [68]:
char_poly_B = det(B - λ * I(3)) # Вычисление характеристического полинома матрицы B

(6 - λ)*((3 - λ)^2)

In [ ]:
println("Характеристический полином матрицы B:\n", char_poly_B) # Вывод характеристического полинома

Характеристический полином матрицы B:
(6 - λ)*((3 - λ)^2)


**Это даст вам точный символьный вид характеристического полинома:**

$$
\det(B - λ I) = (6 - λ)(3 - λ)^2
$$


Мы можем вывести также собственные значния матрицы $B$:

In [70]:
println("Собственные значения матрицы B:\n", eigvals_B) # Вывод собственных значений матрицы B

Собственные значения матрицы B:
[3.0, 3.0, 6.0]


👉 **На этом примере мы изучили:**

- Как использовать символьную матрицу и зачем её задавать.
  
- Как получить аналитические формулы (например, характеристический полином).
  
- Как комбинировать символьный и численный подходы для более глубокого анализа матриц.

<br>

<br>


### 📚 **Дополнительные возможности с `LinearAlgebraX.jl`**

`LinearAlgebraX.jl` - это другой мощный пакет для расширенных матричных вычислений.


#### 🔹 **Пример 4.**

**Выполним установку пакета `LinearAlgebraX.jl`**:

In [40]:
using Pkg
Pkg.add("LinearAlgebraX")

   Resolving package versions...
   Installed IntegerMathUtils ── v0.1.2
   Installed Primes ──────────── v0.5.6
   Installed Mods ────────────── v2.2.6
   Installed Polynomials ─────── v4.0.19
   Installed LinearAlgebraX ──── v0.2.10
   Installed Multisets ───────── v0.4.5
   Installed SimplePolynomials ─ v0.2.18
    Updating `~\Introduction-to-Julia\Project.toml`
  [9b3f67b0] + LinearAlgebraX v0.2.10
    Updating `~\Introduction-to-Julia\Manifest.toml`
  [18e54dd8] + IntegerMathUtils v0.1.2
  [9b3f67b0] + LinearAlgebraX v0.2.10
  [7475f97c] + Mods v2.2.6
  [3b2b4ff1] + Multisets v0.4.5
  [f27b6e38] + Polynomials v4.0.19
  [27ebfcd6] + Primes v0.5.6
  [cc47b68c] + SimplePolynomials v0.2.18
Precompiling project...
   1300.8 ms  ✓ IntegerMathUtils
   1093.3 ms  ✓ Mods
   1125.5 ms  ✓ Multisets
   1033.0 ms  ✓ Primes
  36503.8 ms  ✓ Polynomials
   2394.0 ms  ✓ Polynomials → PolynomialsChainRulesCoreExt
   2542.9 ms  ✓ SimplePolynomials
   2724.4 ms  ✓ Polynomials → PolynomialsMutableArit

In [41]:
using LinearAlgebraX # Подключаем пакет LinearAlgebraX

**Определяем матрицу $A$:**

In [42]:
A = [6 2 1; 0 3 0; 0 0 3] # Создание матрицы A

3×3 Matrix{Int64}:
 6  2  1
 0  3  0
 0  0  3

**Произведём вычисление Жордановой формы:**

In [43]:
J = jordan_form(A) # Получение жордановой формы матрицы A

JordanFactorization{ComplexF64, ComplexF64}(ComplexF64[-0.6666666666666666 + 0.0im -0.3333333333333333 + 0.0im 1.0 + 0.0im; 1.0 + 0.0im 0.0 + 0.0im -0.0 - 0.0im; 0.0 + 0.0im 1.0 + 0.0im -0.0 - 0.0im], ComplexF64[3.0 + 0.0im 0.0 + 0.0im 0.0 + 0.0im; 0.0 + 0.0im 3.0 + 0.0im 0.0 + 0.0im; 0.0 + 0.0im 0.0 + 0.0im 6.0 + 0.0im])

In [44]:
println("Жорданова форма (J):\n", J) # Вывод жордановой формы

Жорданова форма (J):
JordanFactorization{ComplexF64, ComplexF64}(ComplexF64[-0.6666666666666666 + 0.0im -0.3333333333333333 + 0.0im 1.0 + 0.0im; 1.0 + 0.0im 0.0 + 0.0im -0.0 - 0.0im; 0.0 + 0.0im 1.0 + 0.0im -0.0 - 0.0im], ComplexF64[3.0 + 0.0im 0.0 + 0.0im 0.0 + 0.0im; 0.0 + 0.0im 3.0 + 0.0im 0.0 + 0.0im; 0.0 + 0.0im 0.0 + 0.0im 6.0 + 0.0im])


---

<br>

#### 📝 **Заключительные выводы:**

**Факторизация** это мощный инструмент в вычислительной линейной алгебре, делающий сложные вычислительные задачи более простыми и точными. Разные типы факторизаций решают разные задачи, и выбор конкретного подхода зависит от типа исходной матрицы и задачи, которую нужно решить.


⚙️ **Как выбрать нужный тип факторизации?**

Выбор зависит от конкретной задачи и свойств матрицы:

| Свойство матрицы       | Подходящая факторизация        |
|------------------------|--------------------------------|
| Общая матрица          | LU, QR, SVD                    |
| Симметричная положительно определённая | Cholesky, LU, SVD, EVD |
| Неквадратная           | QR, SVD                        |
| Диагонализируемая      | EVD                            |

**Быстрые алгоритмы по сложности** (от простого к сложному):

- **Cholesky** (~$n^3/3$)
  
- **LU** (~$2n^3/3$)
  
- **QR** (~$2n^3$)
  
- **SVD** (~$4n^3$) — наиболее затратный, но универсальный и устойчивый.
  

**Другие менее распространённые факторизации:**

- **Разложение Шура** (Schur Decomposition).
  
- **Разложение Йордана** (Jordan Normal Form).

Эти менее распространённые факторизации широко используются в теоретических исследованиях, а также при решении специфических задач в математике, физике, теории управления и других областях, где необходимы глубокие структурные свойства матриц.

<br>

<br>

---

<br>

<br>

## 🧑‍💻 **3. Специальные матричные структуры.**

<br>

Традиционно считается, что **структура матриц** играет ключевую роль в линейной алгебре.

👉 Чтобы наглядно понять, насколько это важно, рассмотрим систему линейных уравнений с большой матрицей (например, размером 1000×1000):

In [71]:
n = 1000         # Размер матрицы
A = randn(n, n)  # Создание случайной матрицы

1000×1000 Matrix{Float64}:
  0.540699    0.379416     0.0874677  …   1.34572     0.852871   -0.231937
 -0.167392   -0.623696    -0.401732       1.03834     0.889972    0.36893
  0.639121   -0.480105     0.126008      -0.0946948   2.54246     0.497324
  0.0565943   0.513932     0.0510034      0.373774    1.28021    -1.34804
 -0.857251    1.05315     -1.75765       -0.494217    0.145465    0.28542
  1.0764      2.18774      1.28577    …  -0.497476    0.437052    0.316271
 -0.407738    0.504432     0.589993      -2.24159     0.76333     1.75607
 -0.573221   -0.0423707    1.02805       -0.15838     1.18968     1.01841
  0.684799   -0.723471     1.92364       -0.280486   -0.175991    0.147235
 -1.45075     0.663699     0.0526692      1.5899      1.32033     0.367508
  ⋮                                   ⋱                          
 -0.456161    0.713392    -0.0141657      0.61561    -0.363424   -0.0342958
  0.0514155   0.00721383   0.998235      -0.631646   -1.12074     1.32078
  1.16875   

**Julia позволяет проверять свойства матриц, автоматически используя их структуры.**

Например, проверим симметричность матрицы:

In [ ]:
A_sym = A + A'  # Создание симметричной матрицы

issymmetric(A_sym)  # true

true

**Однако иногда численные погрешности могут мешать корректному определению структуры матрицы:**

In [ ]:
A_sym_noisy = copy(A_sym)   # Создаем копию матрицы A_sym
A_sym_noisy[1, 2] += 5eps() # Добавляем шум к элементу матрицы

issymmetric(A_sym_noisy)    # false из-за численной погрешности

false

**К счастью, мы можем явно указать структуру матрицы, используя встроенные типы Julia, такие как:**

- `Diagonal` (диагональная),
  
- `Triangular` (треугольная),
  
- `Symmetric` (симметричная),
  
- `Hermitian` (эрмитова),
  
- `Tridiagonal` (трёхдиагональная),
  
- `SymTridiagonal` (симметричная трёхдиагональная).

Например:

In [ ]:
A_sym_explicit = Symmetric(A_sym_noisy) # Явное приведение к симметричной матрице

1000×1000 Symmetric{Float64, Matrix{Float64}}:
  1.0814      0.212024    0.726589  …   2.4455     -0.883214  -0.921324
  0.212024   -1.24739    -0.881837      1.53908     0.093724   1.93537
  0.726589   -0.881837    0.252015     -0.747585    3.80274   -0.310288
  1.99513     0.0576464   0.850285     -0.0404701   1.25987   -1.8227
 -1.4628      1.53087    -0.960305     -2.86425     1.40123   -0.261048
  0.754237    2.1755      1.44541   …  -1.84052     0.477406   1.27989
  0.005961    0.104977    0.579375     -2.66291    -1.063      0.736079
 -0.453016    1.19606     0.852504      0.130594    1.38034    0.192011
  0.87909    -2.6697      0.888791     -0.294102   -0.631524   0.948341
 -3.01418     0.13447     0.587238      1.64934     0.135734   1.4377
  ⋮                                 ⋱                         
 -0.481482   -0.0124577  -0.901459      0.728234   -0.359166   1.96608
 -0.355685    1.61145     1.87063      -0.680755   -3.21712    0.565884
  0.0132855  -0.759303    1.39975

**Сравним скорость вычисления собственных значений для разных случаев:**

Использование типа `Symmetric` сделает вычисления примерно в 5 раз быстрее!

In [77]:
@time eigvals(A_sym);             # без явного указания структуры
@time eigvals(A_sym_noisy);       # численная погрешность
@time eigvals(A_sym_explicit);    # с явно указанной структурой

  0.099842 seconds (21 allocations: 7.988 MiB, 3.69% gc time)
  0.833271 seconds (27 allocations: 7.928 MiB)
  0.096508 seconds (21 allocations: 7.988 MiB, 3.41% gc time)


<br>

#### 🚩 **Большая проблема и её решение**

Использование специальных типов, таких как `Tridiagonal` и `SymTridiagonal`, позволяет работать с очень большими трёхдиагональными матрицами, которые было бы невозможно хранить и вычислять как обычные (плотные) матрицы.

<br>

**Например, следующую задачу невозможно было бы решить на обычном ноутбуке с плотной матрицей:**

In [78]:
n = 1_000_000    # Размер матрицы
A_large = SymTridiagonal(randn(n), randn(n - 1)) # Создание большой трехдиагональной матрицы

1000000×1000000 SymTridiagonal{Float64, Vector{Float64}}:
 -0.726677   1.45317     ⋅        …   ⋅          ⋅          ⋅ 
  1.45317    0.450956  -1.5558        ⋅          ⋅          ⋅ 
   ⋅        -1.5558     0.451917      ⋅          ⋅          ⋅ 
   ⋅          ⋅        -1.85731       ⋅          ⋅          ⋅ 
   ⋅          ⋅          ⋅            ⋅          ⋅          ⋅ 
   ⋅          ⋅          ⋅        …   ⋅          ⋅          ⋅ 
   ⋅          ⋅          ⋅            ⋅          ⋅          ⋅ 
   ⋅          ⋅          ⋅            ⋅          ⋅          ⋅ 
   ⋅          ⋅          ⋅            ⋅          ⋅          ⋅ 
   ⋅          ⋅          ⋅            ⋅          ⋅          ⋅ 
  ⋮                               ⋱                       
   ⋅          ⋅          ⋅            ⋅          ⋅          ⋅ 
   ⋅          ⋅          ⋅            ⋅          ⋅          ⋅ 
   ⋅          ⋅          ⋅            ⋅          ⋅          ⋅ 
   ⋅          ⋅          ⋅            ⋅          ⋅          ⋅ 
 

In [79]:
@time eigmax(A_large) # Вычисление максимального собственного значения

  1.085284 seconds (620.75 k allocations: 216.017 MiB, 2.69% gc time, 27.10% compilation time)


6.493355010557146

👉 **Благодаря специальным структурам матриц можно эффективно решать масштабные численные задачи, которые иначе были бы слишком ресурсоёмкими для обычных компьютеров.**

<br>

---

<br>

<br>

## ✨ **4. Общая линейная алгебра (Generic linear algebra)**

<br>

Стандартный подход к реализации численной линейной алгебры заключается в использовании подпрограмм из библиотек BLAS и LAPACK. Julia также придерживается этого подхода для матриц с элементами типов `Float32`, `Float64`, `Complex{Float32}` или `Complex{Float64}`.

Однако Julia дополнительно поддерживает **обобщённую (generic)** линейную алгебру. Это позволяет работать с матрицами и векторами, элементами которых являются, например, рациональные числа, комплексные числа произвольной точности и другие типы данных, не обязательно числовые.

<br>

### 🔢 **Рациональные числа (Rational numbers)**

Julia поддерживает рациональные числа из коробки. Рациональные числа можно создавать с помощью двойного слэша `//`:


In [80]:
r = 1//2 # Рациональное число

1//2

**Рациональные числа представляются в виде дроби, что позволяет сохранять точность вычислений, избегая ошибок округления, характерных для чисел с плавающей точкой.**

#### 🔹 **Пример: решение рациональной системы линейных уравнений**

Рассмотрим пример решения системы линейных уравнений, элементы которой — рациональные числа. Чтобы избежать проблем переполнения, характерных при работе с рациональными числами, будем использовать типы большой точности `BigInt`:

- **Создадим матрицу с рациональными элементами типа `BigInt` :**
  
  Генерация случайной рациональной матрицы 3x3

In [85]:
Arational = Matrix{Rational{BigInt}}(rand(1:10, 3, 3)) // 10

3×3 Matrix{Rational{BigInt}}:
 1//5  9//10  4//5
 3//5  4//5   3//10
 3//5  3//10  4//5

- **Создаём вектор $x$, заполненный единицами :**

In [89]:
x = fill(1, 3)

3-element Vector{Int64}:
 1
 1
 1

- **Вычисляем вектор правой части системы :**

In [90]:
b = Arational * x

3-element Vector{Rational{BigInt}}:
 19//10
 17//10
 17//10

- **Решаем систему A * solution = b :**

In [91]:
solution = Arational \ b

3-element Vector{Rational{BigInt}}:
 1
 1
 1

- **Выводим результаты :**

In [92]:
println("Матрица A:\n", Arational)
println("\nВектор b:\n", b)
println("\nРешение системы Ax = b:\n", solution)

Матрица A:
Rational{BigInt}[1//5 9//10 4//5; 3//5 4//5 3//10; 3//5 3//10 4//5]

Вектор b:
Rational{BigInt}[19//10, 17//10, 17//10]

Решение системы Ax = b:
Rational{BigInt}[1, 1, 1]


- **Проверка корректности :**

In [93]:
println("\nПроверка: A * solution == b? ", Arational * solution == b) # true


Проверка: A * solution == b? true


🔴 **Мы можем также явно использовать LU-факторизацию, что полезно при многократном решении системы с одной и той же матрицей :**

Наша **LU-факторизация** матрицы с рациональными элементами будет выглядеть так:

In [94]:
Arational_lu = lu(Arational)

LU{Rational{BigInt}, Matrix{Rational{BigInt}}, Vector{Int64}}
L factor:
3×3 Matrix{Rational{BigInt}}:
  1       0     0
 1//3     1     0
  1    -15//19  1
U factor:
3×3 Matrix{Rational{BigInt}}:
 3//5   4//5    3//10
  0    19//30   7//10
  0      0     20//19

- **Решение с использованием факторизации :**

In [95]:
solution_lu = Arational_lu \ b

3-element Vector{Rational{BigInt}}:
 1
 1
 1

✅ **Оба способа сохраняют точность без перехода к типам с плавающей точкой.**

<br>

#### 🚩 **Преимущества обобщённой линейной алгебры:**

- Возможность работать с широким классом числовых типов, включая точные рациональные и комплексные числа произвольной точности.
  
- Исключение ошибок округления и повышение точности вычислений в теоретических и аналитических задачах.
  
- Удобство и гибкость при работе с математическими и численными задачами различной сложности.

<br>

<br>

---

<br>

### **Упражнения** 🚀 

<br>

#### ✅ Здание 13.1 

Найдите собственные значения матрицы `𝐴`

```julia
A =
[
 140   97   74  168  131
  97  106   89  131   36
  74   89  152  144   71
 168  131  144   54  142
 131   36   71  142   36
]
```

Присвойте результат переменной `A_eigv`.

In [ ]:
using LinearAlgebra

In [ ]:
# Ваше решение:


In [ ]:
@assert A_eigv ==  [-128.49322764802145, -55.887784553056875, 42.7521672793189, 87.16111477514521, 542.4677301466143]

<br>

#### ✅ Здание 13.2 

Создайте диагональную матрицу, состоящую из собственных значений матрицы `A`.

Присвойте результат переменной `A_diag`.

In [ ]:
# Ваше решение:


In [ ]:
@assert A_diag ==  [-128.493    0.0      0.0      0.0       0.0;
    0.0    -55.8878   0.0      0.0       0.0;
    0.0      0.0     42.7522   0.0       0.0;
    0.0      0.0      0.0     87.1611    0.0;
    0.0 0.0      0.0      0.0     542.468]

<br>

#### ✅ Здание 13.3

Создайте нижнюю треугольную матрицу (матрицу, в которой все элементы выше главной диагонали равны нулю) на основе матрицы `A` и сохраните результат в переменной `A_lowertri`.

In [ ]:
# Ваше решение:


In [ ]:
@assert A_lowertri ==  [140    0    0    0   0;
  97  106    0    0   0;
  74   89  152    0   0;
 168  131  144   54   0;
 131   36   71  142  36]